### ISSUE TITLE
Express server breaks in development with HMR

### ISSUE THREAD

**Issue by ZeldOcarina** [remix-run/remix#6283]
### What version of Remix are you using?

1.16.0

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

Create a new project with this package.json:

<details><summary>package.json</summary>
<p>

```javascript
{
  "private": true,
  "sideEffects": false,
  "scripts": {
    "build": "prisma generate && remix build",
    "dev": "remix dev -c \"npm run dev:node\" --port 4001",
    "dev:node": "cross-env NODE_ENV=development nodemon --require dotenv/config ./server.js --watch ./server.js",
    "dev:remix": "remix watch",
    "start": "cross-env NODE_ENV=production node ./server.js",
    "typecheck": "tsc"
  },
  "dependencies": {
    "@aws-sdk/client-s3": "^3.306.0",
    "@aws-sdk/credential-providers": "^3.319.0",
    "@aws-sdk/s3-request-presigner": "^3.319.0",
    "@aws-sdk/util-create-request": "^3.316.0",
    "@aws-sdk/util-format-url": "^3.310.0",
    "@clerk/remix": "^2.5.5",
    "@googlemaps/js-api-loader": "^1.15.1",
    "@popperjs/core": "^2.11.7",
    "@prisma/client": "^4.13.0",
    "@remix-run/css-bundle": "^1.16.0",
    "@remix-run/express": "^1.16.0",
    "@remix-run/node": "^1.16.0",
    "@remix-run/react": "^1.16.0",
    "@types/parse-multipart": "^1.0.0",
    "@types/yaireo__tagify": "^4.17.0",
    "@vercel/node": "^2.6.2",
    "@vercel/remix": "^1.15.0",
    "@yaireo/tagify": "^4.17.7",
    "aws-sdk": "^2.1348.0",
    "axios": "^1.3.6",
    "cross-env": "^7.0.3",
    "html-react-parser": "^3.0.16",
    "isbot": "^3.6.5",
    "npm-run-all": "^4.1.5",
    "parse-multipart": "^1.0.4",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-icons": "^4.8.0",
    "react-phone-input-2": "^2.15.1",
    "react-popper": "^2.3.0",
    "react-spinners": "^0.13.8",
    "remix-utils": "^6.1.0",
    "slugify": "^1.6.6",
    "use-debounce": "^9.0.4"
  },
  "devDependencies": {
    "@remix-run/dev": "^1.15.0",
    "@remix-run/eslint-config": "^1.16.0",
    "@types/react": "^18.0.25",
    "@types/react-dom": "^18.0.8",
    "@types/tinymce": "^4.6.5",
    "eslint": "^8.27.0",
    "nodemon": "^2.0.22",
    "prettier": "2.8.7",
    "prisma": "^4.13.0",
    "tailwindcss": "^3.3.1",
    "typescript": "^4.8.4"
  },
  "engines": {
    "node": ">=14"
  }
}

``` 

</p>
</details> 

Create a new express server with this config:

<details><summary>Express server code</summary>
<p>

```javascript
const path = require("path");

const { broadcastDevReady } = require("@remix-run/node");
const { createRequestHandler } = require("@remix-run/express");
const { installGlobals } = require("@remix-run/node");
const compression = require("compression");
const express = require("express");
const morgan = require("morgan");

installGlobals();

const BUILD_DIR = path.join(process.cwd(), "build");

const app = express();

app.use(compression());

// http://expressjs.com/en/advanced/best-practice-security.html#at-a-minimum-disable-x-powered-by-header
app.disable("x-powered-by");

// Remix fingerprints its assets so we can cache forever.
app.use(
    "/build",
    express.static("public/build", { immutable: true, maxAge: "1y" })
);

// Everything else (like favicon.ico) is cached for an hour. You may want to be
// more aggressive with this caching.
app.use(express.static("public", { maxAge: "1h" }));

app.use(morgan("tiny"));

app.all(
    "*",
    process.env.NODE_ENV === "development"
        ? (req, res, next) => {
            purgeRequireCache();

            return createRequestHandler({
                build: require(BUILD_DIR),
                mode: process.env.NODE_ENV,
            })(req, res, next);
        }
        : createRequestHandler({
            build: require(BUILD_DIR),
            mode: process.env.NODE_ENV,
        })
);
const port = process.env.PORT || 3000;

app.listen(port, () => {
    console.log(`Express server listening on port ${port}`);

    if (process.env.NODE_ENV === 'development') {
        broadcastDevReady(require(BUILD_DIR));
    }
});

function purgeRequireCache() {
    // purge require cache on requests for "server side HMR" this won't let
    // you have in-memory objects between requests in development,
    // alternatively you can set up nodemon/pm2-dev to restart the server on
    // file changes, but then you'll have to reconnect to databases/etc on each
    // change. We prefer the DX of this, so we've included it for you by default
    for (const key in require.cache) {
        if (key.startsWith(BUILD_DIR)) {
            delete require.cache[key];
        }
    }
}
``` 
</p>
</details> 

Run the dev command.

### Expected Behavior

HMR should happen smoothly both on server and client changes.

### Actual Behavior

The app is erroring with:

```
Error: listen EADDRINUSE: address already in use :::3000
    at Server.setupListenHandle [as _listen2] (node:net:1740:16)
    at listenInCluster (node:net:1788:12)
    at Server.listen (node:net:1876:7)
    at Function.listen (/Volumes/T7/web-development.nosync/call-center-directory/node_modules/express/lib/application.js:635:24)
    at Object.<anonymous> (/Volumes/T7/web-development.nosync/call-center-directory/server.js:51:5)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Function.Module._load (node:internal/modules/cjs/loader:958:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
[nodemon] app crashed - waiting for file changes before starting...
```

on all server changes.

**Comment by pcattori** [comment]
This is a bug in the app server restart logic within the dev server.

As a workaround you can use `nodemon --watch build/ server.js` in your `-c` with the `--no-restart` flag:

```json
{
  "scripts": {
    "dev": "remix dev -c 'nodemon --watch build/ server.js'"
  }
}
```

**Comment by pcattori** [comment]
@ZeldOcarina you don't need `dev:remix` anymore. Additionally, you should either:
a) use `--no-restart` + a watcher like `nodemon`: `-c 'nodemon --require dotenv/config --watch build/ ./server.js'`
b) use `node` without `--no-restart`: `-c 'node --require dotenv/config ./server.js'`

You can also omit `NODE_ENV=development` for `remix dev` and any commands run in `-c` (i.e. `dev:node`).
Also the `--port` flag is legacy and doesn't do anything for the new dev server.

In short:

```json
{
  "scripts": {
    "build": "prisma generate && remix build",
    "dev": "remix dev --no-restart -c \"npm run dev:node\"",
    "dev:node": "nodemon --require dotenv/config ./server.js --watch ./server.js",
    "dev:remix": "remix watch",
    "start": "cross-env NODE_ENV=production node ./server.js",
    "typecheck": "tsc"
  },
}
```

**Comment by ZeldOcarina** [comment]
Thanks a lot for the data and the in-depth explanation!
With `--no-restart` though there's no HMR  functionality, so I need to manually refresh.
These are now the current commands:

```javascript
"dev": "remix dev --no-restart -c \"npm run dev:node\"",
"dev:node": "nodemon --require dotenv/config ./server.js --watch ./server.js",
```

**Comment by pcattori** [comment]
If you're using `--no-restart`, you'll need to pick up server changes caused by rebuilds, which means you'll need to also `--watch ./build/` so that your app server restarts when a new `build/index.js` is ready.

**Comment by ZeldOcarina** [comment]
Yeah thanks a lot for the quick fix and for the clarification on the need to use the `--watch` flag on the `./build/`directory. Your patch will be released on the next minor version upgrade right?

**Comment by ZeldOcarina** [comment]
Hi @pcattori!
I've tested the new package by pinning all Remix dependencies @1.16.1-pre.0 and it works beautifully!!
Thanks a lot for the fix! :)
